chain/src/main/scala/org/bitcoins/chain/ChainHandler.scala

```diff
@@ -0,0 +1,65 @@
+package org.bitcoins.chain
+
+import org.bitcoins.chain.api.ChainApi
+import org.bitcoins.chain.blockchain.Blockchain
+import org.bitcoins.chain.blockchain.BlockchainUpdate
+import org.bitcoins.chain.models.{BlockHeaderDAO, BlockHeaderDb}
+import org.bitcoins.core.crypto.DoubleSha256DigestBE
+import org.bitcoins.core.protocol.blockchain.{BlockHeader, ChainParams}
+import org.bitcoins.core.util.BitcoinSLogger
+import org.bitcoins.db.DbConfig
+
+import scala.concurrent.{ExecutionContext, Future}
+
+sealed abstract class ChainHandler extends ChainApi with BitcoinSLogger {
+  def dbConfig: DbConfig
+
+  def chainParams: ChainParams
+
+  implicit def ec: ExecutionContext
+
+  private val blockHeaderDAO = BlockHeaderDAO(chainParams, dbConfig)
+
+  def blockchain: Blockchain
+
+  override def getHeader(
+      hash: DoubleSha256DigestBE): Future[Option[BlockHeaderDb]] = {
+    blockHeaderDAO.findByHash(hash)
+  }
+
+  override def processHeader(header: BlockHeader): Future[ChainHandler] = {
+    val blockchainUpdate = blockchain.connectTip(header)
+
+    val newHandlerF = blockchainUpdate match {
+      case BlockchainUpdate.Successful(blockchain, updatedHeader) =>
+        //now we have successfully connected the header, we need to insert
+        //it into the database
+        val createdF = blockHeaderDAO.create(updatedHeader)
+        createdF.map(_ => ChainHandler(blockchain, chainParams, dbConfig))
+      case BlockchainUpdate.Failed(blockchain, _, reason) =>
+        val errMsg =
+          s"Failed to add header to chain, header=${header.hashBE.hex} reason=${reason}"
+        logger.warn(errMsg)
+        val newHandler = ChainHandler(blockchain, chainParams, dbConfig)
+        Future.successful(newHandler)
+    }
```

Comment:
I'm a bit confused here, in the successful case, it seems that we are treating the database as the source of truth but in the failure case we don't return a failed future. Is there any reason not to return the chain handler right away without waiting on the db write in the success case given handling in the error case?

Reply:
Your right, updated to return a Failed future. Nice catch.